### PR TITLE
Build custom alert message

### PR DESCRIPTION
**Pull request by k-tomoyasu**
Build custom alert messages with query results.  

Alert feature is useful but we cannot see detail about query result.
This feature enable us write custom alert messages with query results (using template engine `Jinja2`).

WIP yet. I would appreciate if  give me feedback.   

alert destination  
- [x] email
- [x] slack 
- [x] webhook
- [x] chatworks
- [x] pagerduty
- [x] mattermost
- [x] hangoutschat
-  ~hipchat~

setting  
setting

slack alert
slack

### CODE REVIEW

**Comment by kocsmy** [comment]
Nice feature @k-tomoyasu — regarding the color of the preview button: we don't use that color so using btn-default would make sense. I'll think about having a better layout for this in the meantime.

**Comment by k-tomoyasu** [comment]
> Nice feature @k-tomoyasu — regarding the color of the preview button: we don't use that color so using btn-default would make sense. I'll think about having a better layout for this in the meantime.

@kocsmy Thanks for your advice! 
Firstly, I changed button color (and some fixes) and update screenshots.

**Comment by k-tomoyasu** [comment]
@jupe Thanks for your comment!
  >  Could rest of integrations be done in follow-up PR?  

That's a big help if it is OK.
I was not good progress of integration because rest destinations are not easy to debug in my environment.

**Comment by jupe** [comment]
For our case email&slack would be enough. Most probably we doesn’t need anything else, at least in near future. Currently slack alert is almost useless because it gives just title, not even link.. so this would be really welcome feature :)

File: client/app/services/alert-template.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+  return Alert;
+}
+
+
+export default function init(ngModule) {
+  ngModule.factory('AlertTemplate', AlertTemplateService);
+
+  ngModule.run(($injector) => {
+    AlertTemplate = $injector.get('AlertTemplate');
```
Comment (rauchy) [inline client/app/services/alert-template.js]: I may be missing something, but why declare this variable in line 3 and assign it here? Why not simply 
```javascript
export const AlertTemplate = $injector.get('AlertTemplate');
```

Comment (rauchy) [reply client/app/services/alert-template.js]: Hehe, disregard this. I was reviewing on a commit-level and saw you already addressed this in a later commit!

File: migrations/versions/ed7bf6adbd4d_add_alert_template.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# revision identifiers, used by Alembic.
+revision = 'ed7bf6adbd4d'
+down_revision = '73beceabb948'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('alerts', sa.Column('template', sa.Text(), nullable=True))
```
Comment (arikfr) [inline migrations/versions/ed7bf6adbd4d_add_alert_template.py]: Let's put `template` in the `options` JSON object we already have in `alerts`.

File: redash/models/__init__.py [hunk 3]
```diff
@@ -772,6 +773,10 @@ def evaluate(self):
     def subscribers(self):
         return User.query.join(AlertSubscription).filter(AlertSubscription.alert == self)
 
+    def render_template(self, showError=None):
```
Comment (arikfr) [inline redash/models/__init__.py]: `showError` seems to be unused? Also should be `show_error`.

**Comment by k-tomoyasu** [comment]
@arikfr Thanks for your review! Sorry for my late reply.

> 1. How about we use Mustache as the template language?  

Agree. I use `Mustache` both server and client.  

> 2. For Email type notification it will be beneficial to be able to edit the subject too.

I think it beneficial. I'll try to consider this feature!

File: redash/models/__init__.py (hunk 3 above)
Comment (k-tomoyasu) [reply redash/models/__init__.py]: thanks! I've removed.

File: migrations/versions/ed7bf6adbd4d_add_alert_template.py (hunk 2 above)
Comment (k-tomoyasu) [reply migrations/versions/ed7bf6adbd4d_add_alert_template.py]: okay, fix logic and remove migration file.

**Comment by k-tomoyasu** [comment]
@arikfr I've improved based on your feedback.

> How about we use Mustache as the template language?

I've modified this.

> For Email type notification it will be beneficial to be able to edit the subject too.

I added `Custom subject` option.  

setting

File: redash/models/__init__.py [hunk 4]
```diff
@@ -772,6 +772,25 @@ def evaluate(self):
...
+            return self.options['template']
+        else:
+            return ""
+
+    @property
+    def custom_subject(self):
+        if 'subject' in self.options:
+            return self.options['subject']
+        else:
+            return ""
```
Comment (arikfr) [inline redash/models/__init__.py]: You can simply call `self.options.get('subject', '')`.

File: redash/models/__init__.py [hunk 5]
```diff
@@ -772,6 +772,25 @@ def evaluate(self):
     def subscribers(self):
         return User.query.join(AlertSubscription).filter(AlertSubscription.alert == self)
 
+    def render_template(self):
+        data = json_loads(self.query_rel.latest_query_data.data)
+        context = {'rows': data['rows'], 'cols': data['columns']}
+        return mustache_render(self.options['template'], context)
```
Comment (arikfr) [inline redash/models/__init__.py]:
```suggestion
        return mustache_render(self.template, context)
```

File: redash/destinations/webhook.py [hunk 6]
```diff
@@ -36,7 +36,8 @@ def notify(self, alert, query, user, new_state, app, host, options):
             data = {
                 'event': 'alert_state_change',
                 'alert': serialize_alert(alert, full=False),
-                'url_base': host 
+                'url_base': host,
+                "description": alert.render_template() if alert.template else ''
```
Comment (arikfr) [inline redash/destinations/webhook.py]: Let's encapsulate this logic in the `render_template` function, i.e. just return an empty string if the template is empty.

File: client/app/pages/alert/index.js [hunk 7]
```diff
@@ -70,6 +76,22 @@ function AlertCtrl($routeParams, $location, $sce, toastr, currentUser, Query, Ev
     );
   };
 
+  this.preview = () => {
+    const notifyError = () => toastr.error('Unable to render description. please confirm your template.', { timeOut: 10000 });
+    try {
+      const result = this.alertTemplate.render(this.alert.options.template, this.queryResult.query_result.data);
+      this.alert.preview = $sce.trustAsHtml(result.escaped);
+      this.alert.previewHTML = $sce.trustAsHtml(result.raw);
```
Comment (arikfr) [inline client/app/pages/alert/index.js]: This needs to be sanitized before trusting.

File: redash/models/__init__.py [hunk 8]
```diff
@@ -772,6 +772,25 @@ def evaluate(self):
     def subscribers(self):
         return User.query.join(AlertSubscription).filter(AlertSubscription.alert == self)
 
+    def render_template(self):
+        data = json_loads(self.query_rel.latest_query_data.data)
+        context = {'rows': data['rows'], 'cols': data['columns']}
```
Comment (arikfr) [inline redash/models/__init__.py]: How about we add to the context information about the alert like the state?

File: redash/models/__init__.py (hunk 4 above)
Comment (k-tomoyasu) [reply redash/models/__init__.py]: thanks. I've fixed.

File: redash/models/__init__.py (hunk 5 above)
Comment (k-tomoyasu) [reply redash/models/__init__.py]: thanks. I've fixed.

File: redash/models/__init__.py (hunk 8 above)
Comment (k-tomoyasu) [reply redash/models/__init__.py]: I added `alert.state` and fixed help message.

File: client/app/pages/alert/index.js (hunk 7 above)
Comment (k-tomoyasu) [reply client/app/pages/alert/index.js]: thank you for pointing mistake!

File: redash/destinations/webhook.py (hunk 6 above)
Comment (k-tomoyasu) [reply redash/destinations/webhook.py]: refactored. thanks!

**Comment by k-tomoyasu** [comment]
@arikfr 

> Thank you for updating this! Please see follow up comments.

Thank you for comments! 
I've updated.  it's ready to review.

**Comment by k-tomoyasu** [comment]
@arikfr
I implemented on other destinations(except `hipchat`).  
(except hipchat).  

> Thank you for updating this! Please see follow up comments.

Thank you for comments!
I've updated. it's ready to review.

**Comment by arikfr** [comment]
Apologies for letting this linger. I will try to get to it this week 🤞

**Comment by buritto** [comment]
Hello, I am also looking forward to this functionality. How soon does this appear in redash?
Thank.

**Comment by arikfr** [comment]
> Apologies for letting this linger. I will try to get to it this week 🤞

"this week"... 2 months later here I am 😅 

To avoid further review cycles, I did some changes myself:

1. Updated with latest master (there were some merge conflicts to fix).
2. Updated how the Webhook destination handles these new options (b078866).
3.  Add log message when checking alert (60cb858, not really related to this PR, just happened to include it).
4. Added a feature flag (`REDASH_FEATURE_EXTENDED_ALERT_OPTIONS`, disabled by default) to control whether to show the new options.

The reason for the feature flag is that the UI still needs more work. I will work on the product and UX definition with @ranbena and we will do these changes ourselves (along with converting this page to React).

But you can already use this in your setup 👍 

Thank you for your effort here, this great feature and patience during the process.

Once the build passes I will merge this.

**Comment by arikfr** [comment]
@k-tomoyasu it's not a big thing, but the email address you signed this commits with (k.tomoyasu@geniee.co.jp) is not associated with your GitHub account. This means that you won't show up in the contributors page on GitHub :-( You can fix this by verifying this email address in your account: https://help.github.com/en/articles/setting-your-commit-email-address.

**Comment by arikfr** [comment]
Merged 👍

**Comment by k-tomoyasu** [comment]
Thanks for fixe and future plan:+1:
I'm looking forward to next release:smile:

> You can fix this by verifying this email address in your account

Thank you for your kindness. But this email is invalid now...
I'm glad to be listed on the page at the next opportunity!

**Comment by Vaidas737** [comment]
Hi, how to enable this feature in version 8.0.0

**Comment by ranbena** [comment]
@vaidas737 https://redash.io/help/user-guide/alerts/custom-alert-notifications

**Comment by Vaidas737** [comment]
Thanks

**Comment by Vaidas737** [comment]
@arikfr thanks a lot

**Comment by auziel** [comment]
> @Vaidas737 in v8, you can enable it with an env var, but note that it won't be 100% compatible with the final implementation in v9.

Hi @arikfr I know it is a year old feature and v9 is already there, but I'm still on v8.0.0. I didn't find a clear explanation how to enable it in v8.
I try to set `FEATURE_EXTENDED_ALERT_OPTIONS=true` which didn't enable the feature. 
Is there a clear doc how to enable this feature for v8.0.0?
How can I customize the message body (I'm using Slack as the alert target)?

**Comment by isdzulqor** [comment]
> > @Vaidas737 in v8, you can enable it with an env var, but note that it won't be 100% compatible with the final implementation in v9.
> 
> Hi @arikfr I know it is a year old feature and v9 is already there, but I'm still on v8.0.0. I didn't find a clear explanation how to enable it in v8.
> I try to set `FEATURE_EXTENDED_ALERT_OPTIONS=true` which didn't enable the feature.
> Is there a clear doc how to enable this feature for v8.0.0?
> How can I customize the message body (I'm using Slack as the alert target)?

try `REDASH_FEATURE_EXTENDED_ALERT_OPTIONS` not `FEATURE_EXTENDED_ALERT_OPTIONS ` @auziel

**Comment by moonsub-kim** [comment]
Can I use this feature with the docker version `redash/redash:10.1.0.b50633`?

**Comment by susodapop** [comment]
> Can I use this feature with the docker version `redash/redash:10.1.0.b50633`?

This feature was replaced in V9 and is now always enabled (the environment variable has no effect). You can read about how to use it [here](https://redash.io/help/user-guide/alerts/custom-alert-notifications).

**Pull request merged**